src/test/java/com/feedzai/commons/sql/abstraction/engine/impl/abs/ForeignKeyTest.java

```diff
@@ -0,0 +1,474 @@
+                .build();
+
+        checkForeignKeys(
+                entity,
+                // USER table contains user 1, but with USER_AUTH_TYPE "LOCAL"; the FK references both columns, and since
+                // their compound values don't match existing values in the referenced table, the entry below should fail
+                entry().set("USER_ID", 1).set("USER_AUTH_TYPE", "EXTERNAL").build(),
+                entry().set("USER_ID", 1).set("USER_AUTH_TYPE", "LOCAL").build(),
+                ImmutableList.of(userEntity)
+        );
+    }
+
+    /**
+     * Tests the usage of a foreign key in a many-to-many relation.
+     *
+     * @throws Exception If something goes wrong in the test.
+     */
+    @Test
+    public void nToNTest() throws Exception {
+        final DbEntity roleEntity = dbEntity()
+                .name("ROLE")
+                .addColumn("ROLE_ID", INT)
+                .pkFields("ROLE_ID")
+                .build();
+
+        engine.addEntity(roleEntity);
+        engine.persist("ROLE", entry().set("ROLE_ID", 1).build());
+
+        // Here we create a table for the relation user-role:
+        // 1 user can have many roles, 1 role may be assigned to many users
+        final DbEntity userRoleEntity = dbEntity()
+                .name("USER_ROLE")
+                .addColumn("USER_ID", INT)
+                .addColumn("ROLE_ID", INT)
+                .addFk(dbFk()
+                                .addColumn("USER_ID")
+                                .referencedTable("USER")
+                                .addReferencedColumn("USER_ID")
+                                .build(),
+                        dbFk()
+                                .addColumn("ROLE_ID")
+                                .referencedTable("ROLE")
+                                .addReferencedColumn("ROLE_ID")
+                                .build()
+                )
+                .pkFields("USER_ID", "ROLE_ID")
+                .build();
+
+        checkForeignKeys(
+                userRoleEntity,
+                // USER table contains user 1, but ROLE table doesn't contain role 2, so the entry below should fail
+                entry().set("USER_ID", 1).set("ROLE_ID", 2).build(),
+                entry().set("USER_ID", 1).set("ROLE_ID", 1).build(),
+                ImmutableList.of(userEntity, roleEntity)
+        );
+    }
+
+    /**
+     * Helper method to verify the behavior of an entity with a foreign key.
+     * <p>
+     * This method performs the following actions:
+     * <ol>
+     *     <li>adds the entity with a foreign key constraint to the database (the primary entities are expected
+     *     to be there already)</li>
+     *     <li>checks that the foreign key constraint prevents the "invalid entry" from being persisted</li>
+     *     <li>persists the "valid entry"</li>
+     *     <li>checks that the foreign key constraint prevents rows referenced by the "valid entry" from being deleted
```

Comment:
Should this be tested here? This depends on the policy used for `on delete` and not the foreign key. Since we don't explicitly set the policy, different DBs might have different defaults and these tests might fail in the future if we add support for other DBs.

This also applies to `removeRowPreviouslyReferencedByForeignKeyTest` and `removeRowAfterRemovingForeignKeyTest`

Reply:
I think it's a good way to test:
- a foreign key constraint should always prevents the "invalid entry" (where the entry references a value that doesn't exist in the reference table) from being persisted, regardless of the ON DELETE/UPDATE setting;
- as strange as it may seem :smile: the DB vendors actually agreed on this one (I don't know if it's explicitly stated on the SQL standard), by default we don't have ON DELETE CASCADE, so deleting a referenced row that is referenced by a foreign key results in an error for all of them; in the future we may add support for that, in which case the tests will need to be updated, they will probably fail